### PR TITLE
Upgrade to hadoop 3.3.5

### PR DESCRIPTION
**Pull request by ayushtkn**
WIP: Exploratory

### What changes were proposed in this pull request?

Upgrade hadoop to 3.3.5

### Why are the changes needed?

To leverage latest hadoop improvements

### Does this PR introduce _any_ user-facing change?

None, that I am aware of.

### How was this patch tested?

CI

### CODE REVIEW

**Comment by aturoczy** [comment]
LGTM, but unfortunately all build are failed. :/

**Comment by ayushtkn** [comment]
Well I have an almost green build, the only test failing is a regular offender, there is already a ticket for that.

Summary of changes:

- Addition of **junit-jupiter-engine** as a dependency in one module: It is coming transitively & now it doesn't come so had to add it.

- **BouncyCastle Upgrade**: Hadoop did that and seems this dependency ain't backward compatible, so we have to upgrade else test fails with NoClassDef Ex.

- The message change in exception: is expired to has expired: **HDFS-16563** changed the exception message, So had to adapt to that.

- Now the interesting part: J**ersey upgrade from 9.4.40.v20210413 to 9.4.45.v20220203**

Why needed:
Hadoop upgraded Jersey and If we don't we get NoClassDef Exceptions
```
java.lang.NoClassDefFoundError: org/eclipse/jetty/server/SymlinkAllowedResourceAliasChecker
	at org.apache.hadoop.http.HttpServer2.addDefaultApps(HttpServer2.java:875)
```
Ref: http://ci.hive.apache.org/job/hive-precommit/job/PR-4290/1/testReport/junit/org.apache.hadoop.hive.ql/TestDDLWithRemoteMetastoreSecondNamenode/Testing___split_01___PostProcess___testCreateDatabaseWithTableNonDefaultNameNode/

Why not moving to 9.4.48.v20220622 which is same as Hadoop version:

- **Hbase Cries**
```
Caused by: java.lang.IllegalArgumentException: headerCacheSize 65536 exceeds maximum value of 65534
	at org.eclipse.jetty.server.HttpConfiguration.setHeaderCacheSize(HttpConfiguration.java:440)
	at org.apache.hadoop.hbase.http.HttpServer$Builder.build(HttpServer.java:384)
	at org.apache.hadoop.hbase.http.InfoServer.<init>(InfoServer.java:87)
	at org.apache.hadoop.hbase.regionserver.HRegionServer.putUpWebUI(HRegionServer.java:2075)
```
So, **the chosen version of Jetty is higher than the current Hive version but lower than the current hadoop 3.3.5 version.**

One solution was to move to Hbase 2.4.0 and leverage [HBASE-23834](https://issues.apache.org/jira/browse/HBASE-23834), but that upgrade doesn't help either, The Jetty exception goes away but the Hbase cluster doesn't start itself. So, playing with Hbase ain't very cool here, We can chase that separately in a different ticket, If it works? Good, If it doesn't work? Try some profiles or something for Hbase, if it still doesn't work. Will start a thread to drop Hbase support from Hive because I have a valid reason "it will block our future Hadoop upgrades". Count my +1(Binding) for that if we reach to that state :) 

The remaining part: Need to try in an actual cluster, to see how things behave in an actual env

File: pom.xml [hunk 1]
```diff
@@ -155,7 +155,7 @@
     <javax-servlet.version>3.1.0</javax-servlet.version>
     <javolution.version>5.5.1</javolution.version>
     <jettison.version>1.5.4</jettison.version>
-    <jetty.version>9.4.48.v20220622</jetty.version>
+    <jetty.version>9.4.45.v20220203</jetty.version>
```
[the remark this answers is not in the document]
Comment (ayushtkn) [reply pom.xml:158]: From Hive point of view it is an upgrade only, but not gonna as high as the one in hadoop 3.3.5, Hbase created some issues. Details in the below summary part.

**Comment by aturoczy** [comment]
> Well I have an almost green build, the only test failing is a regular offender, there is already a ticket for that.
> 
> Summary of changes:
> 
> * Addition of **junit-jupiter-engine** as a dependency in one module: It is coming transitively & now it doesn't come so had to add it.
> * **BouncyCastle Upgrade**: Hadoop did that and seems this dependency ain't backward compatible, so we have to upgrade else test fails with NoClassDef Ex.
> * The message change in exception: is expired to has expired: **HDFS-16563** changed the exception message, So had to adapt to that.
> * Now the interesting part: J**ersey upgrade from 9.4.40.v20210413 to 9.4.45.v20220203**
> 
> Why needed: Hadoop upgraded Jersey and If we don't we get NoClassDef Exceptions
> 
> ```
> java.lang.NoClassDefFoundError: org/eclipse/jetty/server/SymlinkAllowedResourceAliasChecker
> 	at org.apache.hadoop.http.HttpServer2.addDefaultApps(HttpServer2.java:875)
> ```
> 
> Ref: http://ci.hive.apache.org/job/hive-precommit/job/PR-4290/1/testReport/junit/org.apache.hadoop.hive.ql/TestDDLWithRemoteMetastoreSecondNamenode/Testing___split_01___PostProcess___testCreateDatabaseWithTableNonDefaultNameNode/
> 
> Why not moving to 9.4.48.v20220622 which is same as Hadoop version:
> 
> * **Hbase Cries**
> 
> ```
> Caused by: java.lang.IllegalArgumentException: headerCacheSize 65536 exceeds maximum value of 65534
> 	at org.eclipse.jetty.server.HttpConfiguration.setHeaderCacheSize(HttpConfiguration.java:440)
> 	at org.apache.hadoop.hbase.http.HttpServer$Builder.build(HttpServer.java:384)
> 	at org.apache.hadoop.hbase.http.InfoServer.<init>(InfoServer.java:87)
> 	at org.apache.hadoop.hbase.regionserver.HRegionServer.putUpWebUI(HRegionServer.java:2075)
> ```
> 
> So, **the chosen version of Jetty is higher than the current Hive version but lower than the current hadoop 3.3.5 version.**
> 
> One solution was to move to Hbase 2.4.0 and leverage [HBASE-23834](https://issues.apache.org/jira/browse/HBASE-23834), but that upgrade doesn't help either, The Jetty exception goes away but the Hbase cluster doesn't start itself. So, playing with Hbase ain't very cool here, We can chase that separately in a different ticket, If it works? Good, If it doesn't work? Try some profiles or something for Hbase, if it still doesn't work. Will start a thread to drop Hbase support from Hive because I have a valid reason "it will block our future Hadoop upgrades". Count my +1(Binding) for that if we reach to that state :)
> 
> The remaining part: Need to try in an actual cluster, to see how things behave in an actual env

It is not possible to use the 3.0.0-alpha to resolve the issue? I know it is not the best, but as we move ahead with 4.0 or 4.1 in upstream maybe acceptable. I know it is not the best, just an idea. This HBase issue seems pretty problematic.

File: pom.xml (hunk 1 above)
[the remark this answers is not in the document]
Comment (aturoczy) [reply pom.xml:158]: Ack.

**Comment by ayushtkn** [comment]
I tried that as well, try to move to hbase 3.0.0, that is major release for them and offered resistance like anything. The biggest being they removed the entire **hbase-hadoop2-compat** module itself and we rely on that for lot of stuff, not the core of hive though, but still a lot, adapting to that is big work, kept it out from Hadoop upgrade

**Comment by aturoczy** [comment]
Understood :/

**Comment by virajjasani** [comment]
hello, i just came across this based on the thread from Ayush.

> One solution was to move to Hbase 2.4.0 and leverage [HBASE-23834](https://issues.apache.org/jira/browse/HBASE-23834), but that upgrade doesn't help either, The Jetty exception goes away but the Hbase cluster doesn't start itself.

do we have the exception that is stopping hbase? also, would you also once give a try to hbase 2.5.4?

**Comment by virajjasani** [comment]
> The biggest being they removed the entire hbase-hadoop2-compat module itself and we rely on that for lot of stuff

does that mean hive is still using hbase compiled against hadoop 2? is that a mandatory requirement or you would still be fine with hbase 2 compiled against hadoop 3?

also, when you try with hbase 2.5.4, you might want to use `2.5.4-hadoop3` just in case hbase 2 compiled against hadoop 3 works. we use similar in phoenix [here](https://github.com/apache/phoenix/blob/master/pom.xml#L86)

**Comment by ayushtkn** [comment]
Regression in Hadoop, [HADOOP-18724](https://issues.apache.org/jira/browse/HADOOP-18724), need to handle this as well.

WIP

**Comment by virajjasani** [comment]
btw for "hbase unable to get started" related errors, do the errors show up while testing on test cluster or even before that locally only (i.e. while starting miniCluster as part of any UTs)?

**Comment by aturoczy** [comment]
So sad this ticket :(

**Comment by ayushtkn** [comment]
hadoop released 3.3.6, so would chase that @aturoczy

**Pull request closed without merging**